src/expertsystem/reaction/__init__.py

```diff
@@ -429,6 +453,184 @@ def _solve(
+      violations, if its a two body decay. If all initial and final states have
+      the C/G parity defined, then these violations are also determined
+      correctly.
+
+    Args:
+      initial_state: Shortform description of the initial state w/o spin
+        projections.
+      final_state: Shortform description of the final state w/o spin
+        projections.
+
+    Returns:
+      Set of least violating rules. The set can have multiple entries, as
+      several quantum numbers can be violated. Each entry in the tuple
+      represents a group of rules that together violate all possible quantum
+      number configurations.
+    """
+    # pylint: disable=too-many-locals
+    def _check_violations(
+        graph: StateTransitionGraph, node_rules: Set[Rule]
+    ) -> Set[str]:
+        node_id = list(graph.nodes)[0]
+        return validate_fully_initialized_graph(
+            graph,
+            rules_per_node={node_id: node_rules},
+            rules_per_edge={},
+        ).violated_node_rules[node_id]
+
+    # Step 1: create n-body topology
+    # Using and n-body topology is enough, to determine the violations reliably
+    # since only certain spin rules require the isobar model. These spin rules
+    # are not required here though.
+    if not isinstance(initial_state, (list, tuple)):
+        initial_state = [initial_state]  # type: ignore
+    topology_builder = SimpleStateTransitionTopologyBuilder(
+        [
+            InteractionNode(
+                "NBodyScattering", len(initial_state), len(final_state)
+            )
+        ]
+    )
+    topology = topology_builder.build_graphs(
+        len(initial_state), len(final_state)
+    )[0]
+
+    # Step 2: initialize topologies with initial and final state
+    initialized_graphs = initialize_graph(
+        topology=topology,
+        particles=load_default_particles(),
+        initial_state=initial_state,
+        final_state=final_state,
+    )
+
+    # EdgeRules: (can be checked for each node individually)
+    pure_edge_rules: Set[GraphElementRule] = {
+        gellmann_nishijima,
+        isospin_validity,
+    }
+
+    # Step 4: verify edges with the edge rules first
+    edge_check_result = validate_fully_initialized_graph(
+        initialized_graphs[0],
+        rules_per_node={},
+        rules_per_edge={
+            edge_id: pure_edge_rules
+            for edge_id in initialized_graphs[0].get_initial_state_edges()
+            + initialized_graphs[0].get_final_state_edges()
+        },
+    )
+
+    if edge_check_result.violated_edge_rules:
+        raise ValueError(
+            f"Some edges violate"
+            f" {edge_check_result.violated_edge_rules.values()}"
+        )
+
+    # Those rules give the same results, independent on the node and spin props.
+    # Note they are also independent of the topology and hence their results
+    # are always correct.
+    edge_qn_conservation_rules: Set[Rule] = {
```

Comment:
Hm not very DRY, but once again related to #169

Reply:
That's true. I did not have a good and small solutions while coding this. Hence I solved it like this for now. What I can say is that it would be convenient for this check to retrieve rules as bundles, based on their type (edge rule, interaction rule etc.). Also the set of rules is a customized (it not just all rules split up by type).